collab-service/src/history_handler/attempt_cache.ts

```diff
@@ -0,0 +1,127 @@
+import IAttemptCache from './attempt_cache_types';
+import { HistoryAttempt, Question } from '../proto/types';
+import { CreateAttemptResponse } from '../proto/history-crud-service';
+import decodeAttempt from './attempt_decoder';
+
+/**
+ * Intermediate representation of a code submission.
+ * Temporary storage & validator of history attempt.
+ */
+class AttemptCache implements IAttemptCache {
+  question: Question | undefined;
+
+  lang: string;
+
+  submission: string;
+
+  users: string[];
+
+  call: ((attempt: HistoryAttempt) => Promise<CreateAttemptResponse>) | undefined;
```

Comment:
Is this necessary? In my opinion a cache should be basically as dumb as possible, and other logic lifted up to the components that use it.